lib/auth/discoveryconfig/discoveryconfigv1/service.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package discoveryconfigv1
+
+import (
+	"context"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	discoveryconfigv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/discoveryconfig/v1"
+	"github.com/gravitational/teleport/api/types"
+	conv "github.com/gravitational/teleport/api/types/discoveryconfig/convert/v1"
+	"github.com/gravitational/teleport/lib/authz"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// ServiceConfig holds configuration options for the DiscoveryConfig gRPC service.
+type ServiceConfig struct {
+	// Logger is the logger to use.
+	Logger logrus.FieldLogger
+
+	// Authorizer is the authorizer to use.
+	Authorizer authz.Authorizer
+
+	// BackendDiscoveryConfig is the backend for storing DiscoveryConfigs.
+	BackendDiscoveryConfig services.DiscoveryConfigs
+
+	// Clock is the clock.
+	Clock clockwork.Clock
+}
+
+// CheckAndSetDefaults checks the ServiceConfig fields and returns an error if
+// a required param is not provided.
+// Authorizer, Cache and Backend are required params
+func (s *ServiceConfig) CheckAndSetDefaults() error {
+	if s.Authorizer == nil {
+		return trace.BadParameter("authorizer is required")
+	}
+	if s.BackendDiscoveryConfig == nil {
+		return trace.BadParameter("backend for discovery config is required")
+	}
+
+	if s.Logger == nil {
+		s.Logger = logrus.WithField(trace.Component, "DiscoveryConfig.service")
+	}
+
+	if s.Logger == nil {
+		s.Logger = logrus.New().WithField(trace.Component, "discoveryconfig_crud_service")
+	}
```

Comment:
pick one :) looking at all the Component* in constants.go I might name it `discoveryconfig:service`

Reply:
I'll use `"discoveryconfig_crud_service"`
Seems to be used in other places like `access_list`, `okta`, and `user_login_state`